### PR TITLE
Make reconciliation inspectable before execution

### LINKED ISSUE #29: Introduce a pure reconciliation plan
## Problem

`apply` is the only command that produces `SyncOp` values, and those values live in *src/Dojang/Commands/Apply.hs*. `reflect` performs cleanup and copying directly in *src/Dojang/Commands/Reflect.hs*. Conflict detection is also command-specific.

This prevents the three-replica model from answering a basic question as a value: given the observed source, intermediate snapshot, destination, direction, and policy, what should the final replicas be and what operations are required to reach them?

It also forces `apply` to run its source-to-intermediate operations inside `DryRunIO`, observe the simulated result again, and derive destination deletions from that second observation. Destructive effects should be visible in a plan before any interpreter runs.

## Goal

Introduce a reconciliation planner that consumes unified correspondence data and returns a complete, inspectable plan without changing `apply` or `reflect` to use it yet.

## Design

The planner should model at least:

- Direction: source to destination for `apply`, or destination to source for `reflect`.
- Conflict policy: refuse incompatible two-sided changes by default, or prefer the direction's authoritative side when forced.
- Ordered filesystem operations, including directory creation, file copying, file removal, and recursive directory removal.
- Conflicts with enough path and state information for command-specific diagnostics.
- Destructive operations as an explicit subset or classification that callers can inspect before execution.
- No-op and skipped outcomes where needed for diagnostics and tests.

Keep planning separate from execution. Planning should be pure once correspondence and route-state observations are supplied. The executor should be a small `MonadFileSystem` interpreter for the operation vocabulary. Logging, terminal formatting, prompts, hooks, registry updates, and process exit belong outside the core.

The planner should compute desired end states directly. It must not need to mutate a dry-run filesystem and scan it again to determine the second leg of a two-leg synchronization.

The default conflict definition should be directional but symmetric. An incompatible change on both sides relative to the intermediate snapshot is a conflict. A change only on the side that the chosen command will overwrite is not a two-sided conflict. The same rule should hold when source and destination are swapped together with the direction. Cases where both sides independently add equal files or both remove the same tracked entry should converge without a conflict.

Operation order is part of correctness. Plans must create parents before children, remove children before parents, replace incompatible entry kinds safely, and preserve the intermediate recovery copy until the authoritative update has succeeded where failure ordering matters.

## Test strategy

Write planner tests before the implementation. They should use values rather than a real filesystem wherever possible.

Cover at least:

- no changes;
- source-only and destination-only additions, modifications, and removals;
- compatible and incompatible two-sided changes;
- equal independent additions;
- simultaneous removals;
- file-to-directory and directory-to-file replacement;
- symbolic-link states supplied by the unified delta layer;
- forced and unforced policies;
- destructive-operation classification;
- stable operation ordering.

Add properties for directional symmetry, deterministic plans, and the absence of operations for an already reconciled state.

## Scope

A new focused library module is preferred over placing the planner in either command module. `SyncOp` and its ordering should move out of *src/Dojang/Commands/Apply.hs*. The exact module name and type names may follow existing conventions.

This issue may add the executor and its unit tests, but it must not migrate command behavior. Keeping command migration separate makes the planner reviewable without mixing policy design with CLI and hook concerns.

## Completion criteria

- A public or internal library API can produce a complete reconciliation plan from correspondence, direction, and policy.
- Planning performs no filesystem mutation and no terminal or process IO.
- The plan exposes conflicts and destructive operations before execution.
- One `MonadFileSystem` executor interprets the operation vocabulary.
- The planner has table-driven and property-based coverage for direction, conflicts, convergence decisions, and operation ordering.
- `SyncOp` no longer belongs exclusively to `apply`.
- `mise run check` passes.

### PR DESCRIPTION
**Pull request by dahlia**
Closes #29 as the planning layer of #28.

Dojang already observes synchronization through three replicas, but the apply path could only discover its full effect by executing part of the work in `DryRunIO` and scanning the simulated filesystem again.  That made deletion guards and conflict handling properties of a command pipeline instead of values that callers could inspect before mutation.

This PR draws a strict boundary between observation, planning, and execution. The effectful boundary turns a `FileCorrespondence` and route state into a `ReconciliationInput`.  From there, the planner is pure: it takes a direction and conflict policy, chooses the authoritative replica, and returns desired end states together with conflicts, skipped entries, destructive operations, and a stable operation sequence.  A small `MonadFileSystem` interpreter executes that sequence without taking on command-layer concerns such as prompts, hooks, logging, or exit codes.

The plan models source/intermediate/destination convergence directly rather than simulating its first leg.  Incompatible changes on both outer replicas are reported as conflicts, while equal independent additions and simultaneous removals converge without one.  Refusing conflicts blocks the whole plan before the executor mutates anything.  A forced policy keeps the same analysis but selects the authoritative side explicitly.

Operation order carries the recovery guarantees that were previously implicit in command code.  Parent directories are created before their children, recursive removals absorb redundant descendant work, and incompatible entry kinds are removed before replacement.  Deletions remove the replica being overwritten before discarding the intermediate copy, so a failed first step does not destroy the remaining recovery state.

`SyncOp` and its interpreter now live in *src/Dojang/Types/Reconciliation.hs*.  *src/Dojang/Commands/Apply.hs* calls the shared interpreter, but still uses its existing planner and command behavior. Keeping migration out of this PR makes the planning policy reviewable on its own and leaves the apply/reflect transition to the next sub-issue.

The tests use table-driven examples for boundary cases and Hedgehog properties for the rules that should hold across arbitrary states.  They check directional symmetry, determinism under reordered inputs, convergence after execution, and idempotence on a second plan.  Focused examples cover conflicts, equal additions, simultaneous removals, file/directory replacement, symbolic links, ignored destinations, destructive-operation classification, and recovery-safe ordering.  Executor tests also verify that a refused plan performs no partial mutation.

### CODE REVIEW

**Comment by dahlia** [comment]
@codex review

File: src/Dojang/Types/Reconciliation.hs [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+          FileEntry correspondence.intermediate.path authoritative.stat
+        targetOperations =
+          if not targetNeedsUpdate
+            then []
+            else
+              tagOperations SecondPhase overwrittenReplica $
+                replaceEntry stagedIntermediate overwritten
+        operations = intermediateOperations ++ targetOperations
+    in (if null operations then NoChange else WillReconcile, operations)
+
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/Dojang/Types/Reconciliation.hs:415]: Addressed in 301a103: ignored apply targets now suppress destination operations while source-to-intermediate reconciliation still proceeds, with regression coverage.

File: test/Dojang/Types/ReconciliationSpec.hs [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+    (modelDelta intermediateState destinationState)
+    (modelComparison sourceState destinationState)
+    routeState
+
+
+applyModelOperations
+  :: Map OsPath ModelState
+  -> [PlannedSyncOp]
+  -> Map OsPath ModelState
+applyModelOperations = foldl' applyOperation
```
[the remark this answers is not in the document]
Comment (dahlia) [reply test/Dojang/Types/ReconciliationSpec.hs:124]: Declined: this module builds and all 329 tests pass under the project's GHC 9.10.3 toolchain. In this version, Prelude exports foldl', so an explicit Data.List import is not required.

**Comment by dahlia** [comment]
@codex review

File: src/Dojang/Types/Reconciliation.hs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+            | otherwise -> WillReconcile
+    in (outcome, operations)
+
+  planExisting :: (ReconciliationOutcome, [TaggedOperation])
+  planExisting =
+    let ignored = case (direction, input.destinationRouteState) of
+          (SourceToDestination, Ignored route pattern)
+            | targetNeedsUpdate ->
+                Just $ IgnoredDestination route pattern
+          _ -> Nothing
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/Dojang/Types/Reconciliation.hs:391]: Addressed in 5d01357: reflection now skips plans whose authoritative destination is ignored, with regression coverage for a destination-only addition.

**Comment by dahlia** [comment]
@codex review
